common/lockmanager_test.go

```diff
@@ -0,0 +1,362 @@
+
+    key1 := "testkey1";
+
+    Lock(key1);
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+}
+
+// Lock(1) -> Unlock(1) -> !Unlock(1).
+func TestUnlockingAnUnlockedLock(test *testing.T) {
+    clear();
+    defer clear();
+
+    key1 := "testkey1";
+
+    Lock(key1);
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+
+    err = Unlock(key1);
+    if (err == nil) {
+        test.Fatalf("Failed to return an error when unlocking a lock that is already unlocked.");
+    }
+}
+
+// Unlock(1).
+func TestUnlockingKeyThatDoesntExist(test *testing.T) {
+    clear();
+    defer clear();
+
+    doesNotExistKey := "dne";
+
+    err := Unlock(doesNotExistKey);
+    if (err == nil) {
+        test.Fatalf("Failed to return an error when unlocking a key that does not exist.");
+    }
+}
+
+// Lock(1) -> %Lock(2) -> Unlock(1) -> %Unlock(2).
+func TestMainThreadUnlocksFirst(test *testing.T) {
+    clear();
+    defer clear();
+
+    key1 := "testkey1";
+    key2 := "testkey2";
+    var key1UnlockBlock sync.WaitGroup;
+    var key2LockedBlock sync.WaitGroup;
+    var threadFinishBlock sync.WaitGroup;
+
+    Lock(key1);
+
+    threadFinishBlock.Add(1);
+    key2LockedBlock.Add(1);
+    key1UnlockBlock.Add(1);
+
+    go func() {
+        defer threadFinishBlock.Done();
+
+        // Lock key2 while key1 is already locked.
+        Lock(key2);
+
+        key2LockedBlock.Done();
+
+        // Wait for the main thread to unlock key1.
+        key1UnlockBlock.Wait();
+
+        err := Unlock(key2);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock key2 in the child thread.");
+        }
+    }();
+
+    // Wait for the thread to lock key2 before unlocking key1.
+    key2LockedBlock.Wait();
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock key1 in the main thread.");
+    }
+
+    key1UnlockBlock.Done();
+
+    // Wait for the thread to finish unlocking key2.
+    threadFinishBlock.Wait();
+}
+
+// Lock(1) -> %Lock(2) -> %Unlock(2) -> Unlock(1).
+func TestChildThreadUnlocksFirst(test *testing.T) {
+    clear();
+    defer clear();
+
+    key1 := "testkey1";
+    key2 := "testkey2";
+    var key2LockUnlockBlock sync.WaitGroup;
+
+    Lock(key1);
+
+    key2LockUnlockBlock.Add(1);
+
+    go func() {
+        defer key2LockUnlockBlock.Done();
+
+        // Lock key2 while key1 is already locked.
+        Lock(key2);
+
+        err := Unlock(key2);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock key2 in the child thread.");
+        }
+    }();
+
+    // Wait for the thread to lock and unlock key2 before unlocking key1.
+    key2LockUnlockBlock.Wait();
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock key1 in the main thread.");
+    }
+}
+
+// Lock(1) -> %Lock(1) -> Unlock(1) -> %Unlock(1).
+func TestMainThreadUnlockFirst(test *testing.T) {
```

Comment:
This name is too close to "TestMainThreadUnlocksFirst" (note the "s").
Either one can change their name, or both.

In this situation, I would probably scope each name with something like "OneLock"/"SingleLock" and "TwoLocks"/"DoubleLocks".
You can do it like that, or however you like.